### DISCUSSION TITLE
A separate typing package?

### DISCUSSION THREAD

**Discussion by lucascolley** [data-apis/array-api#863] [General]
jumping off from https://github.com/data-apis/array-api/pull/857#issuecomment-2499147994 @asmeurer 

---

For various reasons discussed in the linked issue, it seems like this repo may not be the ideal location for the development of static typing for the standard. One proposed idea is to create a new package to house type stubs, which could be used by array libraries to check their compliance, and be used to type libraries which consume array API standard arrays.

One idea which seems to have agreement:

- The stubs in this repo (i.e. the signatures, English descriptions of types, and English descriptions of behaviour) should remain the ultimate source of truth. The 'proper' type stubs should try to faithfully replicate the standard (but perhaps may not be able to get all the way there due to upstream Python typing limitations).

Otherwise, does a separate repo sound like a good idea? Or would it fit into one of the existing `array-api-*` repos? If we want to do this, how should we go about it?

Also quoting some higher-level questions from @jorenham:

> - Will it be used as a verification suite for the static type annotations of array-api libraries?
> - Will there be a "baseline" package that these protocols should at least be compatible with, e.g. numpy?
> - Should it follow the official python typing specification, and if so, how will that be verified?
> - Will it be distributed as a (standalone- or bundled-) package, and if so, should these types also be usable at runtime?

ping @betatim @rgommers @34j @BvB93 @nstarman @vnmabus @pmeier @tomnicholas

**Comment by nstarman** [comment]
This sounds like a great idea. I'm going to ping @ntessore, who's been thinking about this as well.

I appreciate having a single source of truth and if the consensus is to have that be the narrative docs, that sounds good to me.
Also having an official PyPI-installable library of runtime-checkable Protocol objects would be fantastic.

Some thoughts:

Such a package would presumably define

- `Array` (possibly parametrized)
- `Namespace` for the API namespace
- `DType`
- `Device`
- etc.

IMO this would not be in `array-api-strict`, etc, as it is would be an implementation-free definition of the Array API. I do think there is also space for a  "universal" library that will work with any compatible arrays, without vendoring one itself. This could be in the typing package, or done separately. For example `cos` would be implemented as:

```python
import array_api as api

def cos(x: api.Array, ...) -> api.Array:
    xp = x.__array_namespace__()
    return xp.cos(x)
```

A few years back I wrote an example library that defined the Protocols and the "universal" functions: https://github.com/nstarman/array_api

**Reply by 34j** [reply]
> A few years back I wrote an example library that defined the Protocols and the "universal" functions: [nstarman/array_api](https://github.com/nstarman/array_api)

Wow I don't know which version of array-api your code is based on, but it is fantasitic in that it is fully checked and formatted. I think it is better to start here than with the code in this repository.

> ```python
> import array_api as api
> 
> def cos(x: api.Array, ...) -> api.Array:
>     xp = x.__array_namespace__()
>     return xp.cos(x)
> ```

However, I think that Generic should be used for Array type (as @jorenham mentioned here https://github.com/data-apis/array-api/pull/857#issuecomment-2499121020
(For both the code in the package and the code of the end-user.)
(Because this code does not eliminate the possibility of `torch.Tensor` being returned when `np.ndarray` is passed, for example, as `api.Array` seems to be a Protocol)

**Reply by ntessore** [reply]
Thanks for the ping @nstarman! Indeed, we have an ongoing project to make our cosmological simulation code fully Array API-compliant and typing is a central part of that. I would be happy for us to contribute to a coordinated effort instead of building our own solution.

As for making Array generic, I guess the idea would be to write functions as accepting an `ArrayT` typevar having `api.Array` as an upper bound. So typing library might provide an `api.ArrayT`, but the `api.Array` as such doesn't need to be generic. (Unless this is about parametrising the array type, e.g. `np.floating`, which I'm not sure about.)

**Reply by nstarman** [reply]
> Wow I don't know which version of array-api your code is based on, but it is fantasitic in that it is fully checked and formatted. I think it is better to start here than with the code in this repository.

I would be happy to transfer that repo to the data-apis org if people think it's a good starting point.

**Reply by rgommers** [reply]
> I would be happy to transfer that repo to the data-apis org if people think it's a good starting point.

Since the outcome of this discussion thread + the community meeting discussion we just had is to indeed create a new `array-api-typing` package with at least you and @jorenham as maintainers, I'd say if you both agree that that is a good starting point then sure let's do that. Or I can create a fresh repo and add you to it - just let me know!

**Reply by jorenham** [reply]
How about a fresh repo where we can gradually port over the relevant bits of https://github.com/nstarman/array_api? That way we can start by setting up the project and the tooling to validate the types against existing array-api libraries, and then incrementally port over the stuff from https://github.com/nstarman/array_api, while ensuring compatibility with the ecosystem.

**Reply by rgommers** [reply]
All thumbs up, so let's go! I created https://github.com/data-apis/array-api-typing and @data-apis/typing-team with maintain permissions. In addition, I gave @jorenham and @nstarman full admin access (setting up CI etc. will likely need that). Happy to help if other things are needed of course. If we want to give more folks commit rights (now or in the future) then please ping me - I'd say we let @jorenham and @nstarman make those decisions for now, so that they can really own the repo/package.

For populating the new repo, it'd be great if you copy over the license and code of conduct metadata files from this repo.

**Reply by 34j** [reply]
@jorenham @nstarman @rgommers 
Unfortunately, development seems to have stopped at the part of setting up the development environment.

Could you please give me full access to the repository as well? Because:

1. I maintain a Copier template for Python packages (https://github.com/browniebroke/pypackage-template) and can instantly (like 5 minutes) set up the development environment and upload packages to PyPI automatically as well as host documentation (, which requires access to settings). This template should cover most of @nstarman 's proposal (https://github.com/data-apis/array-api-typing/issues/1) and changes can be made later easily.
2. I already have most of the Protocol auto-generation scripts completed in #858, and I think I can *automatically* generate Protocols with Device, Dtype, and Shape typing at least for specific operations (elementwise etc.). The maintenance status of the repository really suggests that this automation is very important and, by its very nature, should be implemented from the beginning.

Specifically, I would like to have complete control of the repository for a week or so, and the opportunity to complete the first release (or first PR to be reviewed). If there are any change requests after that, I will comply fully and not interfere. I really just want this feature to be completed.

**Reply by jorenham** [reply]
> Unfortunately, development seems to have stopped at the part of setting up the development environment.

Not "halted", just postponed. I plan to continue working on it once I have the bandwidth to do so.

> Could you please give me full access to the repository as well?

Why not just submit a PR? I wouldn't mind reviewing it.

**Reply by 34j** [reply]
You have a point, but it has already been 5 months since the repository stopped activity and seems to be stuck in configuring development environment (seems there's no discussion at all). it may be possible with PR, but full access is eventually required to set up a development environment, and I think I am very familiar with it. I believe this is relatively a simple task that should not take years to complete or should be waited until the bandwidth is available, and I would like to help "kick starting" the repository.

**Reply by jorenham** [reply]
I'll welcome all the help that we can get. So how about we start with a PR for the dev setup, and then we'll see from there?

**Reply by 34j** [reply]
It is not impossible, but it is very tedious as it requires me to instruct you every step of the way which requires admin access. It's just applying the above template; it would look like https://github.com/34j/jordan-form . What is the reason I cannot be the maintainer though suggesting improvements a lot here in the first place? Because I'm not an array api member?

**Reply by nstarman** [reply]
Agreed. It's a package we all need but have not had the requisite time to actively work on.

**Reply by jorenham** [reply]
> What is the reason I cannot be the maintainer though suggesting improvements a lot here in the first place? 

Because we don't know whether your vision aligns with ours. But even if you were given commit rights, then we still would want to see a PR, because we don't self-merge.

**Reply by 34j** [reply]
I have explained many times here as well as in #858 what I am about to do. If there is a chance that my entire proposal could be kicked because it does not "match your vision", then of course I cannot make such an effort. And I believe "no self-merge" will not make sense in setup phase.
Anyway if that is the case, I hope the work by you is completed soon.

**Reply by 34j** [reply]
I am still not comfortable with the idea of continuing a one-sided and asymmetrical relationship and censorship despite almost no progress for 5 or 6 months. Since it is stuck on the setup phase, there should be nothing to lose. I am not trying to control everything, I just want it to be completed. I hope you will reconsider and could collaborate with me.😉

**Reply by NeilGirdhar** [reply]
Just chiming in as an interested outsider.  I’d really love to see progress on Array API typing, and it seems like there’s momentum and willingness from all sides.

I can understand @34j’s frustration with PRs left hanging or discarded, especially when a lot of work has been put in. That said, @jorenham has been super responsive and thorough in past reviews.  Why not start with a small PR to test the waters?  And then progressively do larger PRs as you build trust and momentum?

Really excited to see this take shape!

**Reply by 34j** [reply]
@NeilGirdhar I did my best, but it was impossible. The maintainers spent 3 days complaining about [my PR](https://github.com/data-apis/array-api-typing/pull/14) and then merged their “preferred” setup only in [3 hours](https://github.com/data-apis/array-api-typing/pull/12) while I was sleeping after I gave up, which means I seem to be somewhat slowing down the development. However, I am happy that my abnormal behavior has allowed the development to resume.

Anyway my work is already done with https://github.com/34j/array-api. You can enjoy array-api-compat typing instantly by:

 ```shell
pip install array-api
```

![image](https://github.com/user-attachments/assets/20f56bf2-02fa-4e57-8bfb-a39d7fc9f339)

![image](https://github.com/user-attachments/assets/3eadf245-ad8e-44d3-ac76-9e1d75049037)

as well as `ArrayNamespace` and `ArrayNamespaceFull`.

See https://github.com/data-apis/array-api/discussions/951 for further explanation about its core feature, the *automated* typing code generation.

**Reply by nstarman** [reply]
Very cool stuff @34j!

Yes, it's a shame that you were unwilling to go through the standard review process of pull requests and engage in the standard practices of an open source community. Instead of being ashamed by the speedy PR process of https://github.com/data-apis/array-api-typing/pull/12, https://github.com/data-apis/array-api-typing/pull/16, and https://github.com/data-apis/array-api-typing/pull/17 I would encourage you to see those for what they are: that people are happy to engage with and review small, well crafted PRs, especially when they are motivated by discussion from Issues: https://github.com/data-apis/array-api-typing/issues/1, https://github.com/data-apis/array-api-typing/issues/6, https://github.com/data-apis/array-api-typing/issues/8.
The 3 days "complaining" about your PR were related to the large number of cruft files, which I'll note are still present in your new repo: e.g. 1 of many https://github.com/34j/array-api/blob/main/src/array_api/main.py.

As you noted:
> @nstarman Yes, I understand what you are saying, but you are being too particular. I don't care about the details the setup. If I were in your shoes, I would want to complain massively about my PR, and I agree with you 100%:

So please take the correct message from this process. Open source work happens in a community. Clearly that community is capable of moving quickly to review PRs that are atomic and when the tone is genial, rather than grumpy. Good features are appreciated. Your PRs are welcome.

On another note about etiquette in an open-source community, you'll note that I had a package — https://github.com/nstarman/array_api — that's quite old but which I never registered on pypi. That's because camping on the name `array_api` when I'm not associated with the official consortium would be rude.

**Reply by nstarman** [reply]
@34j I see a number of edits to your comment. To address the current iteration: "seem to be somewhat slowing down the development. However, I am happy that my abnormal behavior has allowed the development to resume."

You'll find that @jorenham has repeatedly requested major changes to my PRs (and vice versa). We engage in discussion to find the best solution. As stated, we would be very happy for you to contribute to development, which you were not slowing, but speeding up, as you note. The only requirement is a willingness to discuss and adapt. No dictators. No self-merge. Consensus by hopefully polite conversation.
In particular, your notes about automated code generation of typing stubs were and are well received! The automated generation of unneeded config files less so.

**Reply by rgommers** [reply]
Hi all, thanks for trying to work through this. I'm sure we'd all like to both see progress on static typing support, and work together constructively. @NeilGirdhar's [comment above](https://github.com/data-apis/array-api/discussions/863#discussioncomment-13305661) about how to start contributing seems like very good advice to me. Let me also post a link to an blog post that seems to capture a lot of the dynamics here: https://gael-varoquaux.info/programming/technical-discussions-are-hard-a-few-tips.html.

I'll also acknowledge here that some of the earliest static typing work, by @nstarman, was stalled because I (and others) found it hard to make enough time to review. As a result we collectively decided to create `array-api-typing` and trust @nstarman and @jorenham as the initial maintainers. Contributing PRs that are well-scoped will help them and build trust with them. And hopefully the static typing team will grow over time.

@34j thank you for the work you're doing on improving static typing. I hope you'll try again to contribute some of that to `array-api-typing`. Creating a competing package that may be more feature-complete, have a larger scope, or is different in some other way is also perfectly fine - competition can be healthy and serve different needs. I'd just respectfully ask to not name your package `array-api`, because that looks like the "core package name" for the whole project and the thing that `array-api-compat`, `array-api-strict`, and `array-api-extra` are built around. We discussed this thread in the community meeting yesterday, and there were some concerns from other participants there (who didn't comment on this thread) as well. If you could choose a different name for your package and give the `array-api` name to the consortium for future usage, we'd really appreciate it.

**Reply by NeilGirdhar** [reply]
@34j 

Sorry I didn't reply sooner, and thank you for taking the time to draft a well-thought-out pull request. I know the outcome was frustrating, and I really appreciate the effort you put in.

I wanted to share a few personal reflections from earlier in my career that might resonate, but feel free to skip them if they don’t.

Many years ago, during an internship at Nvidia, I stayed up late one night to prepare a solution to a tricky problem. I presented it to the team and was thrilled when they accepted my approach.

Later that term, I noticed that my manager working hard on something I thought could be done much more simply. I couldn’t convince him to switch, so I stayed after hours and quietly built the simpler version. It worked well—faster and more elegant—but when I showed it to him, he wasn’t happy. He was annoyed. I thought I had helped. But my dad said something that stuck with me: “_If you're not making your manager happy, why even go in to work?_”

That experience taught me that good ideas aren’t always enough: we also have to consider timing, process, and relationships. Communication matters. And sometimes, making space for others' contributions matters more than having the “best” solution.

More broadly, I’ve come to see how easily our egos get entangled with our work. When things go well, it’s tempting to say _I did this_. When they don’t, it becomes _I failed_. That kind of thinking is seductive, but isolating and ultimately Sisyphean. What’s helped me most when faced with disagreements is refocusing on the collective effort: _We did this_.

We all want `array-api-typing` to move forward, but we're all volunteering, and none of us wants to do all of the work alone.  That's the point of collaboration.  I get that it can feel draining when you have a clear plan in mind, but others seem to be in the way.  Still, teamwork (when it works) isn't just more efficient.  It's deeply rewarding.  Nothing gives our work (and our lives) more meaning than the relationships we build.

**Comment by 34j** [comment]
In addition to the above matters, I suppose that the stub package should manually maintain only the classes that a TypeVar would frequently bound (`Array`, `Dtype`, `Device`, `.... .info`) , and the `Protocol`s for other array-api functions should be automatically generated by a script. The following script https://github.com/data-apis/array-api/pull/858/files#diff-e86d9c4a581472cb53f58194cf98904663bc563f864a9e29edbd181dd12eec5d modified from https://github.com/data-apis/array-api/pull/685#issuecomment-1720065338 can be used as a prototype.

Typing functions as `Protocol`s with dtype as a `TypeVar` could be complicated, but roughly speaking, there should be only two types of functions: creation functions and functions that handle numerical/any values, therefore I would like propose sticking to automatic generation.

**Comment by ntessore** [comment]
As a very happy adopter of an Array API typing spec, should one come to exist, here's my take on the questions you raise:

> The stubs in this repo (i.e. the signatures, English descriptions of types, and English descriptions of behaviour) should remain the ultimate source of truth. The 'proper' type stubs should try to faithfully replicate the standard (but perhaps may not be able to get all the way there due to upstream Python typing limitations).

Fully agreed.

> Otherwise, does a separate repo sound like a good idea? Or would it fit into one of the existing array-api-* repos? If we want to do this, how should we go about it?

As a user I would prefer a separate repo, if only to illustrate the distinction between the actual standard, the strict implementation, the compatibility layer, and the typing implementation.

> Will there be a "baseline" package that these protocols should at least be compatible with, e.g. numpy?

I think that would defeat the purpose. It's easy to test that my library works with numpy (use numpy in testing). It's hard to verify my library works with any compliant array package.

> Should it follow the official python typing specification, and if so, how will that be verified?

I'm not sure what that means, exactly, but you might want to pick one of the type checkers and verify that a test suite of annotations passes without complaints.

> Will it be distributed as a (standalone- or bundled-) package, and if so, should these types also be usable at runtime?

I would very much hope for a stand-alone package that users of my library don't even have to worry about, unless we are in a `TYPE_CHECKING` situation.

**Reply by jorenham** [reply]
> > Should it follow the official python typing specification, and if so, how will that be verified?
> 
> I'm not sure what that means, exactly, but you might want to pick one of the type checkers and verify that a test suite of annotations passes without complaints.

This is what I meant: https://typing.readthedocs.io/en/latest/spec/index.html
And as far as static-type checkers are concerned; Pyright is a lot better at following it than mypy ([source](https://github.com/erictraut/mypy_issues)). An even better option would be [basedpyright](https://github.com/detachhead/basedpyright/), which is a fork of pyright that has more (stricter) rules and better defaults.

**Reply by jorenham** [reply]
> > Will it be distributed as a (standalone- or bundled-) package, and if so, should these types also be usable at runtime?
> 
> I would very much hope for a stand-alone package that users of my library don't even have to worry about, unless we are in a `TYPE_CHECKING` situation.

The reason I was asking, is because typing in python isn't limited to static typing, as there are also many examples of applications of runtime typing (e.g. [pydantic](https://github.com/pydantic/pydantic), [typer](https://github.com/fastapi/typer), [beartype])(https://github.com/beartype/beartype), [optype](https://github.com/jorenham/optype), etc). And in this case, the "runtime-protocols", which are `typing.Protocol` types that are decorated with `@typing.runtime_checkable`, might be interesting for the `array-api-typing` package.

**Comment by lucascolley** [comment]
@nstarman just to split this to a separate thread;

> I do think there is also space for a "universal" library that will work with any compatible arrays, without vendoring one itself.

There are two similar efforts here. Xarray has implemented something similar for standard arrays wrapped by Xarray: https://github.com/pydata/xarray/pull/9776.

There is also https://github.com/data-apis/array-api-extra, for “universal” functions which are not in the standard.

I suppose I don’t see the need for the kind of library you suggest. Is the main idea to avoid consumers having to call `__array_namespace__` and keep track of `xp`? If so, that seems useful for new libraries, although I think it may be a bit too implicit for converting existing libraries - I think explicitly passing around `xp` when support is partial can make it clear what is supposed to be working.

**Reply by nstarman** [reply]
> I suppose I don’t see the need for the kind of library you suggest.

Not a high priority. It just enables very easy support for all implementing array types. Converting a program that only supports numpy can become as simple as changing `import numpy as np` to `import array_api as np`.

**Reply by nstarman** [reply]
This might also be a good idea for a "array_api_compat.universal" library that eliminates even the need to call `array_api_compat.array_namespace"...

**Reply by lucascolley** [reply]
I think the main reason that hasn't happened yet it that it means `array_namespace` is called for every function call rather than just once.

**Reply by 34j** [reply]
https://github.com/ivy-llc/ivy is a "universal" library I suppose (although it is not FOSS)

**Comment by jorenham** [comment]
I was planning on building this myself within [`optype`](https://github.com/jorenham/optype/), so that I could use it in [`scipy-stubs`](https://github.com/jorenham/scipy-stubs). But it's also fine by me if we make it into a separate package. 

And for what it's worth; in https://github.com/jorenham/optype/issues/25 I explain how you could go about typing the "untypeable" dtypes and devices.
